ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs

```diff
@@ -301,6 +305,269 @@ prop_diffusion_target_established_local defaultBearerInfo diffScript =
+-- but for running on Diffusion. This means it has to have in consideration the
+-- the logs for all nodes running will all appear in the trace and the test
+-- property should only be valid while a given node is up and running.
+--
+-- We do not need separate above and below variants of this property since it
+-- is not possible to exceed the target.
+--
+prop_diffusion_target_active_below :: AbsBearerInfo
+                                   -> DiffusionScript
+                                   -> Property
+prop_diffusion_target_active_below defaultBearerInfo diffScript =
+    let sim :: forall s . IOSim s Void
+        sim = diffusionSimulation (toBearerInfo defaultBearerInfo)
+                                  diffScript
+                                  tracersExtraWithTimeName
+                                  tracerDiffusionSimWithTimeName
+
+        events :: [Events DiffusionTestTrace]
+        events = fmap ( Signal.eventsFromList
+                      . fmap (\(WithName _ (WithTime t b)) -> (t, b))
+                      )
+               . Trace.toList
+               . splitWithNameTrace
+               . Trace.fromList ()
+               . fmap snd
+               . Signal.eventsToList
+               . Signal.eventsFromListUpToTime (Time (10 * 60 * 60))
+               . Trace.toList
+               . fmap (\(WithTime t (WithName name b)) -> (t, WithName name (WithTime t b)))
+               . withTimeNameTraceEvents
+                  @DiffusionTestTrace
+                  @NtNAddr
+               $ runSimTrace sim
+
+     in conjoin
+      $ verify_target_active_below
+      <$> events
+
+  where
+    verify_target_active_below :: Events DiffusionTestTrace -> Property
+    verify_target_active_below events =
+      let govLocalRootPeersSig :: Signal (LocalRootPeers.LocalRootPeers NtNAddr)
+          govLocalRootPeersSig =
+            selectDiffusionPeerSelectionState Governor.localRootPeers events
+
+          govEstablishedPeersSig :: Signal (Set NtNAddr)
+          govEstablishedPeersSig =
+            selectDiffusionPeerSelectionState
+              (EstablishedPeers.toSet . Governor.establishedPeers)
+              events
+
+          govActivePeersSig :: Signal (Set NtNAddr)
+          govActivePeersSig =
+            selectDiffusionPeerSelectionState Governor.activePeers events
+
+          govActiveFailuresSig :: Signal (Set NtNAddr)
+          govActiveFailuresSig =
+              Signal.keyedLinger
+                180 -- 3 minutes  -- TODO: too eager to reconnect?
+                (fromMaybe Set.empty)
+            . Signal.fromEvents
+            . Signal.selectEvents
+                (\case TracePromoteWarmFailed _ _ peer _ ->
+                         --TODO: the environment does not yet cause this to happen
+                         -- it requires synchronous failure in the establish
+                         -- action
+                         Just (Set.singleton peer)
+                       --TODO
```

Comment:
todo what?

Reply:
Good question, since I just copy pasted the same code from https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network/test/Test/Ouroboros/Network/PeerSelection.hs#L1544 I thought it would be worth to preserve the TODOs as well.